File: packages/shared/src/evaluator.js
export const EvaluationReason = Object.freeze({
  FLAG_NOT_FOUND: "FLAG_NOT_FOUND",
  FLAG_DISABLED: "FLAG_DISABLED",
  MATCHED_RULE: "MATCHED_RULE",
  PERCENTAGE_ROLLOUT: "PERCENTAGE_ROLLOUT",
  DEFAULT_RULE: "DEFAULT_RULE"
});

export function evaluateFlag(flag, user, options = {}) {
  const defaultValue = options.defaultValue ?? false;

  if (!flag) {
    return {
      flagKey: options.flagKey,
      enabled: defaultValue,
      reason: EvaluationReason.FLAG_NOT_FOUND
    };
  }

  if (!flag.enabled) {
    return {
      flagKey: flag.key,
      enabled: false,
      reason: EvaluationReason.FLAG_DISABLED
    };
  }

  const matchedRule = firstMatchedRule(flag.targetingRules ?? [], user ?? {});
  if (matchedRule) {
    return {
      flagKey: flag.key,
      enabled: true,
      reason: EvaluationReason.MATCHED_RULE,
      matchedRule
    };
  }

  const rolloutPercentage = normalizeRolloutPercentage(flag.rolloutPercentage ?? 0);
  if (rolloutPercentage > 0) {
    const bucket = bucketUser(flag.key, user?.id);
    return {
      flagKey: flag.key,
      enabled: bucket < rolloutPercentage,
      reason: EvaluationReason.PERCENTAGE_ROLLOUT,
      bucket,
      rolloutPercentage
    };
  }

  return {
    flagKey: flag.key,
    enabled: defaultValue,
    reason: EvaluationReason.DEFAULT_RULE
  };
}

export function firstMatchedRule(rules, user) {
  return rules.find((rule) => matchesRule(rule, user));
}

export function matchesRule(rule, user) {
  const actual = user?.[rule.attribute];
  if (actual === undefined || actual === null) {
    return false;
  }

  switch (rule.operator) {
    case "equals":
      return actual === rule.value;
    case "not_equals":
      return actual !== rule.value;
    case "contains":
      return String(actual).includes(String(rule.value));
    case "ends_with":
      return String(actual).endsWith(String(rule.value));
    case "in":
      return Array.isArray(rule.value) && rule.value.includes(actual);
    default:
      return false;
  }
}

export function bucketUser(flagKey, userId) {
  if (!userId) {
    return 99;
  }

  return stableHash(`${flagKey}:${userId}`) % 100;
}

export function stableHash(input) {
  let hash = 2166136261;

  for (let index = 0; index < input.length; index += 1) {
    hash ^= input.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }

  return hash >>> 0;
}

function normalizeRolloutPercentage(value) {
  if (!Number.isFinite(value)) {
    return 0;
  }

  return Math.max(0, Math.min(100, Math.floor(value)));
}

File: services/config-service/internal/flags/evaluator.go
package flags

import "strings"

const (
	ReasonFlagNotFound      = "FLAG_NOT_FOUND"
	ReasonFlagDisabled      = "FLAG_DISABLED"
	ReasonMatchedRule       = "MATCHED_RULE"
	ReasonPercentageRollout = "PERCENTAGE_ROLLOUT"
	ReasonDefaultRule       = "DEFAULT_RULE"
)

func Evaluate(flag *Flag, flagKey string, user map[string]any, defaultValue bool) EvaluateFlagResponse {
	if flag == nil {
		return EvaluateFlagResponse{
			FlagKey: flagKey,
			Enabled: defaultValue,
			Reason:  ReasonFlagNotFound,
		}
	}

	if !flag.Enabled {
		return EvaluateFlagResponse{
			FlagKey: flag.Key,
			Enabled: false,
			Reason:  ReasonFlagDisabled,
		}
	}

	if len(flag.TargetingRules) > 0 {
		matchedRule := firstMatchedRule(flag.TargetingRules, user)
		if matchedRule != nil {
			return EvaluateFlagResponse{
				FlagKey: flag.Key,
				Enabled: true,
				Reason:  ReasonMatchedRule,
			}
		}
	}

	if flag.RolloutPercentage > 0 {
		userID, ok := user["id"].(string)
		if !ok || userID == "" {
			return EvaluateFlagResponse{
				FlagKey: flag.Key,
				Enabled: defaultValue,
				Reason:  ReasonDefaultRule,
			}
		}

		bucket := bucketUser(flag.Key, userID)
		rolloutPercenteage := flag.RolloutPercentage

		return EvaluateFlagResponse{
			FlagKey:           flag.Key,
			Enabled:           bucket < flag.RolloutPercentage,
			Reason:            ReasonPercentageRollout,
			Bucket:            &bucket,
			RolloutPercentage: &rolloutPercenteage,
		}
	}

	if len(flag.TargetingRules) == 0 {
		return EvaluateFlagResponse{
			FlagKey: flag.Key,
			Enabled: true,
			Reason:  ReasonDefaultRule,
		}
	}

	return EvaluateFlagResponse{
		FlagKey: flag.Key,
		Enabled: defaultValue,
		Reason:  ReasonDefaultRule,
	}
}

func firstMatchedRule(rules []TargetingRule, user map[string]any) *TargetingRule {
	for _, rule := range rules {
		if matchesRule(rule, user) {
			return &rule
		}
	}

	return nil
}

func matchesRule(rule TargetingRule, user map[string]any) bool {
	actual, exists := user[rule.Attribute]
	if !exists || actual == nil {
		return false
	}

	switch rule.Operator {
	case "equals":
		return actual == rule.Value
	case "not_equals":
		return actual != rule.Value
	case "contains":
		actualString, ok := actual.(string)
		if !ok {
			return false
		}

		valueString, ok := rule.Value.(string)
		if !ok {
			return false
		}

		return strings.Contains(actualString, valueString)
	case "ends_with":
		actualString, ok := actual.(string)
		if !ok {
			return false
		}

		valueString, ok := rule.Value.(string)
		if !ok {
			return false
		}

		return strings.HasSuffix(actualString, valueString)
	default:
		return false
	}
}

func bucketUser(flagKey string, userID string) int {
	hash := fnv32a(flagKey + ":" + userID)
	return int(hash % 100)
}

func fnv32a(input string) uint32 {
	var hash uint32 = 2166136261

	for _, char := range input {
		hash ^= uint32(char)
		hash *= 16777619
	}

	return hash
}
